ts/src/pro/xt.ts

```diff
@@ -0,0 +1,1062 @@
+                //            "refreshToken": "eyJhbGciOiqJSUzI1NiJ9.eyJhY2NvdW50SWQiOiIyMTQ2Mjg1MzIyNTU5Iiwic3ViIjoibGh4dDRfMDAwMUBzbmFwbWFpbC5jYyIsInNjb3BlIjoicmVmcmVzaCIsImlzcyI6Inh0LmNvbSIsImxhc3RBdXRoVGltZSI6MTY2MzgxMzY5MDk1NSwic2lnblR5cGUiOiJBSyIsInVzZXJOYW1lIjoibGh4dDRfMDAwMUBzbmFwbWFpbC5jYyIsImV4cCI6MTY2NjQwNTY5MCwiZGV2aWNlIjoidW5rbm93biIsInVzZXJJZCI6MjE0NjI4NTMyMjU1OX0.Fs3YVm5YrEOzzYOSQYETSmt9iwxUHBovh2u73liv1hLUec683WGfktA_s28gMk4NCpZKFeQWFii623FvdfNoteXR0v1yZ2519uNvNndtuZICDdv3BQ4wzW1wIHZa1skxFfqvsDnGdXpjqu9UFSbtHwxprxeYfnxChNk4ssei430"
+                //        }
+                //    }
+                //
+                const result = this.safeValue (response, 'result');
+                client.subscriptions['accessToken'] = this.safeString (result, 'accessToken');
+            }
+        }
+        return client.subscriptions['accessToken'];
+    }
+
+    async subscribe (name: string, access: string, methodName: string, market: object = undefined, params = {}) {
+        /**
+         * @ignore
+         * @method
+         * @description Connects to a websocket channel
+         * @see https://doc.xt.com/#websocket_privaterequestFormat
+         * @see https://doc.xt.com/#futures_market_websocket_v2base
+         * @param {string} name name of the channel
+         * @param {string} access public or private
+         * @param {string} methodName the name of the CCXT class method
+         * @param {object|undefined} market CCXT market
+         * @param {object} params extra parameters specific to the xt api
+         * @returns {object} data from the websocket stream
+         */
+        await this.loadMarkets ();
+        const privateAccess = access === 'private';
+        let isContract = undefined;
+        [ isContract, params ] = this.handleIsContractAndParams (methodName, market, params);
+        const subscribe = {
+            'method': isContract ? 'SUBSCRIBE' : 'subscribe',
+            'id': this.numberToString (this.milliseconds ()) + name,  // call back ID
+        };
+        if (privateAccess) {
+            if (!isContract) {
+                subscribe['params'] = [ name ];
+                subscribe['listenKey'] = await this.getListenKey (isContract);
+            } else {
+                const listenKey = await this.getListenKey (isContract);
+                const param = name + '@' + listenKey;
+                subscribe['params'] = [ param ];
+            }
+        } else {
+            subscribe['params'] = [ name ];
+        }
+        const tradeType = isContract ? 'contract' : 'spot';
+        const messageHash = name + ':' + tradeType;
```

Comment:
@samgermain right now you're not using the symbol to build `messageHash`, so how would you correctly resolve two different futures, targeting different markets?

```Typescript
a = await xt.watchTrades('BTC/USDT')
b = await xt.watchTrades('LTC/USDT')
```
⬆️ if you get an update from the `LTC/USDT` market how do you resolve only `b`'s future?

Reply:
> @samgermain right now you're not using the symbol to build `messageHash`, so how would you correctly resolve two different futures, targeting different markets?
> 
> ```ts
> a = await xt.watchTrades('BTC/USDT')
> b = await xt.watchTrades('LTC/USDT')
> ```
> 
> ⬆️ if you get an update from the `LTC/USDT` market how do you resolve only `b`'s future?

The marketId is included in the name, and is present as part of the name in the response

Reply:
> > @samgermain right now you're not using the symbol to build `messageHash`, so how would you correctly resolve two different futures, targeting different markets?
> > ```ts
> > a = await xt.watchTrades('BTC/USDT')
> > b = await xt.watchTrades('LTC/USDT')
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ⬆️ if you get an update from the `LTC/USDT` market how do you resolve only `b`'s future?
> 
> The marketId is included in the name, and is present as part of the name in the response

@samgermain I'm not saying you cannot subscribe properly, I'm saying you can't resolve the futures properly after getting an update, have you tried to do it ?

Reply:
@samgermain I'm not sure I understand what you mean, I'm talking about resolving the futures indexed by the messagehash. If inside client.futures you have only `ticker:spot` how can you resolve it correctly if you have the example I posted above?

Reply:
look at [lines like this one,](https://github.com/ccxt/ccxt/pull/17814/files#r1206100758) the market id is part of the name

Reply:
@samgermain oh I see it now, a bit unconventional so I got confused sorry